sympy/series/order.py

```diff
@@ -220,74 +220,70 @@ def __new__(cls, expr, *args, **kwargs):
-                    expr = expr.as_independent(*args, as_Add=False)[1]
-
-                    expr = expand_power_base(expr)
-                    expr = expand_log(expr)
-
-                    if len(args) == 1:
-                        # The definition of O(f(x)) symbol explicitly stated that
-                        # the argument of f(x) is irrelevant.  That's why we can
-                        # combine some power exponents (only "on top" of the
-                        # expression tree for f(x)), e.g.:
-                        # x**p * (-x)**q -> x**(p+q) for real p, q.
-                        x = args[0]
-                        margs = list(Mul.make_args(
-                            expr.as_independent(x, as_Add=False)[1]))
-
-                        for i, t in enumerate(margs):
-                            if t.is_Pow:
-                                b, q = t.args
-                                if b in (x, -x) and q.is_real and not q.has(x):
-                                    margs[i] = x**q
-                                elif b.is_Pow and not b.exp.has(x):
-                                    b, r = b.args
-                                    if b in (x, -x) and r.is_real:
-                                        margs[i] = x**(r*q)
-                                elif b.is_Mul and b.args[0] is S.NegativeOne:
-                                    b = -b
-                                    if b.is_Pow and not b.exp.has(x):
-                                        b, r = b.args
-                                        if b in (x, -x) and r.is_real:
-                                            margs[i] = x**(r*q)
-
-                        expr = Mul(*margs)
+                from sympy import PoleError, Function
+                try:
+                    expr = expr.as_leading_term(*args)
+                except PoleError:
+                    if isinstance(expr, Function) or\
+                            all(isinstance(arg, Function) for arg in expr.args):
+                        # It is not possible to simplify an expression
+                        # containing only functions (which raise error on
+                        # call to leading term) further
+                        pass
+                    else:
+                        orders = []
+                        pts = tuple(zip(args, ps))
+                        for arg in expr.args:
+                            try:
+                                lt = arg.as_leading_term(*args)
+                            except PoleError:
+                                lt = arg
+                            if lt not in args:
+                                order = Order(lt)
+                            else:
+                                order = Order(lt, *pts)
+                            orders.append(order)
+                        if expr.is_Add:
+                            new_expr = Order(Add(*orders), *pts)
```

Comment:
The order of a sum cannot in general be derived from the orders of its arguments. The leading terms of the arguments may cancel each other, and then a series expansion has to be used to find the first non-canceled term. That is what is done by `Add._eval_as_leading_term`. The same procedure is needed here.

Reply:
Ohh okay that maybe the case , I haven't changed much of the core code yet and this patch was also used originally . I'm not sure ,will have to go through the debug again but I feel this has been taken care of . This is what is happening in the issue referenced above.
```
>>> Order(1 - cos(1/x), (x, oo))
O(x**(-2), (x, oo))
>>> Order(1 - 1/(2*x**2) - cos(1/x), (x, oo))
O(x**(-4), (x, oo)) 
>>> Order(sin(1/x)*x - cos(1/x), (x, oo)) # here 1 cancels out and the second term of the cos and sin terms
O(x**(-2), (x, oo))               # is correctly evaluated to give the following result
```
This wasn't working before as `extract_leading_term()` was not functioning as required. Could you give some examples where you feel this could go wrong

Reply:
Are there examples where this line (247) is executed?

Reply:
Not anything particular that I could think of, off the topic of my head . Not sure it's required or not cause I hadn't changed this patch as it was originally used. Will check some examples from the tests !